### PR TITLE
move from c++14 to c++17

### PR DESCRIPTION
**Pull request by ObiWahn**

### CODE REVIEW

File: 3rdParty/iresearch/core/index/field_data.cpp [hunk 1]
```diff
@@ -696,7 +696,7 @@ void fields_data::flush(field_writer& fw, flush_state& state) {
 
   {
     static struct less_t {
```
Comment (ObiWahn) [inline 3rdParty/iresearch/core/index/field_data.cpp]: Why not a lambda?

File: 3rdParty/iresearch/core/index/segment_writer.cpp [hunk 2]
```diff
@@ -125,7 +125,7 @@ bool segment_writer::flush(std::string& filename, segment_meta& meta) {
   // flush columnstore and columns indices
   if (col_writer_->flush() && !columns_.empty()) {
     static struct less_t {
```
Comment (ObiWahn) [inline 3rdParty/iresearch/core/index/segment_writer.cpp]: Why not a lambda?

**Comment by jsteemann** [comment]
What about changing the following places as well?
```
3rdParty/linenoise-ng/CMakeLists.txt:  set(BASE_COMPILER_OPTIONS "-std=c++14 -Wall -Wextra")
3rdParty/linenoise-ng/CMakeLists.txt:  set(BASE_COMPILER_OPTIONS "-std=c++14 -Wall -Wextra")
3rdParty/linenoise-ng/CMakeLists.txt:  set(BASE_COMPILER_OPTIONS "-std=c++14 -Wall -Wextra")
3rdParty/V8/v5.7.492.77/gypfiles/standalone.gypi:              'CLANG_CXX_LANGUAGE_STANDARD': 'c++14',  # -std=c++14
3rdParty/rocksdb/v5.16.X/CMakeLists.txt:  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++14")
3rdParty/V8/v5.7.492.77/gypfiles/standalone.gypi:          '-std=gnu++14',
3rdParty/V8/v5.7.492.77/gypfiles/standalone.gypi:          '-std=gnu++14',
3rdParty/V8/v5.7.492.77/gypfiles/standalone.gypi:                       '-std=gnu++14' ],
3rdParty/V8/v5.7.492.77/gypfiles/standalone.gypi:          'cflags_cc': [ '-std=gnu++14', ],
```

**Comment by ObiWahn** [comment]
We do not need to modify 3rdParty libs. As long as we use the same compiler the ABI will match.

**Comment by jsteemann** [comment]
@ObiWahn : although it is true that the ABI may match, it can be a problem if the libraries compile in different features based on the c++ standard and also expose them in their header files.
At least ICU does so:
```
3rdParty/V8/v5.7.492.77/third_party/icu/source/common/unicode/platform.h:#elif !defined(__cplusplus)
3rdParty/V8/v5.7.492.77/third_party/icu/source/common/unicode/platform.h:#elif __cplusplus >= 201402L
3rdParty/V8/v5.7.492.77/third_party/icu/source/common/unicode/platform.h:#elif __cplusplus >= 201103L
3rdParty/V8/v5.7.492.77/third_party/icu/source/common/unicode/platform.h:#ifdef __cplusplus
```
So it's unsafe to not make the change throughout the libraries we use.

**Comment by ObiWahn** [comment]
Could you please explain the problem a bit more in detail? I do not understand why a lib that is compiled with standard a should not be include-able by a lib that uses a newer standard? For example in the distribution I use there are several libs that do not require c++17 and I happily use their interfaces. If not only internal details change due to a newer standard, but the interface (e.g. types) the library offers I would consider it a bug. Mixing pre 11 and later versions could be afaik dangerous when using GCC, but that should not be true for a mix of 11,14 and 17. Anyway I'll add the changes, later to eliminate your doubts.

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1059/

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/2066/

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1184/

**Pull request merged**